File: src/initializing_pattern_iterator.cc
#include "initializing_pattern_iterator.h"

#include "gpio.h"
#include "wiring_pi_wrapper.h"
#include "enqueued_pattern_iterator.h"
#include "table_constants.h"
#include "sisyphus_util.h"
#include "model.pb.h"

namespace {

EnqueuedPatternIterator ParametricToIterator(const sisyphus::StoredPattern& pattern) {
  return EnqueuedPatternIterator(SisyphusUtil::StoredPatternToPattern(pattern));
}

// Spiral circling from inside out
EnqueuedPatternIterator SweepPattern() {
  sisyphus::StoredPattern stored_pattern;
  auto* pattern = stored_pattern.mutable_parametric();
  pattern->set_start_t(0);
  pattern->set_end_t(5000);
  pattern->set_t_step(1);
  pattern->set_type(sisyphus::ParametricPattern::POLAR);

  auto* theta = pattern->mutable_equation_for_x_or_theta()->mutable_binary_equation();
  theta->set_operation(sisyphus::ParametricPattern_Equation_BinaryEquation::DIVIDE);
  theta->mutable_equation_left()->set_variable(sisyphus::ParametricPattern_Equation::T);
  theta->mutable_equation_right()->set_constant(15);

  auto* r = pattern->mutable_equation_for_y_or_r()->mutable_binary_equation();
  r->set_operation(sisyphus::ParametricPattern_Equation_BinaryEquation::MULTIPLY);
  r->mutable_equation_left()->set_variable(sisyphus::ParametricPattern_Equation::T);
  // 0.95 * .0002
  r->mutable_equation_right()->set_constant(0.00019);

  return ParametricToIterator(stored_pattern);
}

// Spiral circling from outside in
EnqueuedPatternIterator TargetPattern() {
  sisyphus::StoredPattern stored_pattern;
  auto* pattern = stored_pattern.mutable_parametric();
  pattern->set_start_t(0);
  pattern->set_end_t(5000);
  pattern->set_t_step(1);
  pattern->set_type(sisyphus::ParametricPattern::POLAR);

  auto* theta = pattern->mutable_equation_for_x_or_theta()->mutable_binary_equation();
  theta->set_operation(sisyphus::ParametricPattern_Equation_BinaryEquation::DIVIDE);
  theta->mutable_equation_left()->set_variable(sisyphus::ParametricPattern_Equation::T);
  theta->mutable_equation_right()->set_constant(-15);

  auto* r = pattern->mutable_equation_for_y_or_r()->mutable_binary_equation();
  r->set_operation(sisyphus::ParametricPattern_Equation_BinaryEquation::SUBTRACT);
  r->mutable_equation_left()->set_constant(0.95);
  auto* sub_r = r->mutable_equation_right()->mutable_binary_equation();
  sub_r->set_operation(sisyphus::ParametricPattern_Equation_BinaryEquation::MULTIPLY);
  sub_r->mutable_equation_left()->set_variable(sisyphus::ParametricPattern_Equation::T);
  // 0.95 * .0002
  sub_r->mutable_equation_right()->set_constant(0.00019);

  return ParametricToIterator(stored_pattern);
}

EnqueuedPatternIterator sweep_out = SweepPattern();
EnqueuedPatternIterator sweep_in = TargetPattern();

}

InitializingPatternIterator::InitializingPatternIterator(const sisyphus::Pattern& pattern)
    : stage(INITIALIZING_STAGE_ZEROING), current(0), complete(false),
    zero_count(0), center_steps(0) {
  if (pattern.path_segment().size() < 1) {
    return;
  }
  target = pattern.path_segment(0).start().linear_value();
}

bool InitializingPatternIterator::has_next() const {
  return complete;
}

const sisyphus::Step& InitializingPatternIterator::next() {
  sisyphus::Step step;
  switch (stage) {
    case INITIALIZING_STAGE_ZEROING:
      if (Gpio::read(GPIO_LINEAR_FEEDBACK) == LOW) {
        zero_count = 0;
        step.set_linear_movement(sisyphus::Step::BACKWARDS);
      } else {
        if (++zero_count == ZEROING_THRESHOLD) {
          stage = INITIALIZING_STAGE_CENTERING;
        }
      }
      break;

    case INITIALIZING_STAGE_CENTERING:
      if (center_steps <= LINEAR_CENTER_OFFSET_STEPS) {
        step.set_linear_movement(sisyphus::Step::FORWARDS);
        center_steps++;
      } else {
        stage = INITIALIZING_STAGE_CLEARING;
      }
      break;

    case INITIALIZING_STAGE_CLEARING:
      if (sweep_out.has_next()) {
        const sisyphus::Step& s = sweep_out.next();
        step.set_linear_movement(s.linear_movement());
        step.set_angular_movement(s.angular_movement());
        if (s.linear_movement() == sisyphus::Step::FORWARDS) {
          current++;
        }
      } else {
        stage = INITIALIZING_STAGE_TARGETING;
      }
      break;

    case INITIALIZING_STAGE_TARGETING:
      if (!sweep_out.has_next()) {
        complete = true;
        break;
      }
      const sisyphus::Step& s = sweep_out.next();
      step.set_linear_movement(s.linear_movement());
      step.set_angular_movement(s.angular_movement());
      if (s.linear_movement() == sisyphus::Step::BACKWARDS) {
        current--;
      }
      if (current <= target) {
        complete = true;
      }
      break;
  }
  return std::move(step);
}

bool InitializingPatternIterator::is_external_pattern() const {
  return false;
}

File: node/initializing_pattern_iterator.js
const AbstractPatternIterator = require('./abstract_pattern_iterator');
const EnqueuedPatternIterator = require('./enqueued_pattern_iterator');
const Model = require('./gen/model_pb');
const Gpio = require('./gpio');
const TableConstants = require('./table_constants');
const SisyphusUtil = require('./sisyphus_util');

const INITIALIZING_STAGE_ZEROING = 1;
const INITIALIZING_STAGE_CENTERING = 2;
const INITIALIZING_STAGE_CLEARING = 3;
const INITIALIZING_STAGE_TARGETING = 4;

const ZEROING_THRESHOLD = 100;

function parametricToIterator(pattern) {
  return new EnqueuedPatternIterator(SisyphusUtil.storedPatternToPattern(pattern));
}

function binaryEquation(eq) {
  const equation = new Model.ParametricPattern.Equation();
  equation.setBinaryEquation(eq);
  return equation;
}

function variable() {
  const variable = new Model.ParametricPattern.Equation();
  variable.setVariable(Model.ParametricPattern.Equation.Variable.T);
  return variable;
}

function constant(constantVal) {
  const constant = new Model.ParametricPattern.Equation();
  constant.setConstant(constantVal);
  return constant;
}

// Spiral circling from inside out
function sweepPattern() {
  const storedPattern = new Model.StoredPattern();
  const parametricPattern = new Model.ParametricPattern();
  parametricPattern.setStartT(0);
  parametricPattern.setEndT(5000);
  parametricPattern.setTStep(1);
  parametricPattern.setType(Model.ParametricPattern.Type.POLAR);

  const theta = new Model.ParametricPattern.Equation.BinaryEquation();
  theta.setOperation(Model.ParametricPattern.Equation.BinaryEquation.BinaryEquationType.DIVIDE);
  theta.setEquationLeft(variable());
  theta.setEquationRight(constant(20));
  parametricPattern.setEquationForXOrTheta(binaryEquation(theta));

  const r = new Model.ParametricPattern.Equation.BinaryEquation();
  r.setOperation(Model.ParametricPattern.Equation.BinaryEquation.BinaryEquationType.MULTIPLY);
  r.setEquationLeft(variable());
  r.setEquationRight(constant(0.00019));
  parametricPattern.setEquationForYOrR(binaryEquation(r));

  storedPattern.setParametric(parametricPattern);

  return parametricToIterator(storedPattern);
}

// Spiral circling from outside in
function targetPattern() {
  const storedPattern = new Model.StoredPattern();
  const parametricPattern = new Model.ParametricPattern();
  parametricPattern.setStartT(0);
  parametricPattern.setEndT(5000);
  parametricPattern.setTStep(1);
  parametricPattern.setType(Model.ParametricPattern.Type.POLAR);

  const theta = new Model.ParametricPattern.Equation.BinaryEquation();
  theta.setOperation(Model.ParametricPattern.Equation.BinaryEquation.BinaryEquationType.DIVIDE);
  theta.setEquationLeft(variable());
  theta.setEquationRight(constant(-20));
  parametricPattern.setEquationForXOrTheta(binaryEquation(theta));

  const r = new Model.ParametricPattern.Equation.BinaryEquation();
  r.setOperation(Model.ParametricPattern.Equation.BinaryEquation.BinaryEquationType.SUBTRACT);
  r.setEquationLeft(constant(0.95));
  const subR = new Model.ParametricPattern.Equation.BinaryEquation();
  subR.setOperation(Model.ParametricPattern.Equation.BinaryEquation.BinaryEquationType.MULTIPLY);
  subR.setEquationLeft(variable());
  subR.setEquationRight(constant(0.00019));
  r.setEquationRight(subR);
  parametricPattern.setEquationForYOrR(binaryEquation(r));

  storedPattern.setParametric(parametricPattern);

  return parametricToIterator(storedPattern);
}

class InitializingPatternIterator extends AbstractPatternIterator {
  constructor(pattern) {
    this.target = 0;
    this.current = 0;
    this.stage = INITIALIZING_STAGE_ZEROING;
    this.zeroCount = 0;
    this.centerSteps = 0;
    this.complete = false;
    this.clearingPattern = new EnqueuedPatternIterator(sweepPattern());
    this.targetingPattern = new EnqueuedPatternIterator(targetPattern());

    if (pattern.getPathSegmentList().length < 1) {
      return;
    }
    this.clearingPattern.reset();
    this.targetingPattern.reset();
    this.target = pattern.getPathSegmentList()[0].getStart().getLinearValue();
  }

  hasNext() {
    return !this.complete;
  }

  next() {
    const step = new Model.Step();
    step.setAngularMovement(Model.Step.Movement.STOP);
    step.setLinearMovement(Model.Step.Movement.STOP);

    switch (this.stage) {
      case INITIALIZING_STAGE_ZEROING:
        if (Gpio.read(Gpio.GPIO_LINEAR_FEEDBACK) == Gpio.LOW) {
          this.zeroCount = 0;
          step.setLinearMovement(Model.Step.Movement.BACKWARDS);
        } else if (++this.zeroCount == ZEROING_THRESHOLD) {
          console.log("CENTERING");
          this.stage = INITIALIZING_STAGE_CENTERING;
        }
        break;
      case INITIALIZING_STAGE_CENTERING:
        if (this.centerSteps <= TableConstants.LINEAR_CENTER_OFFSET_STEPS) {
          step.setLinearMovement(Model.Step.Movement.FORWARDS);
          this.centerSteps++;
        } else {
          console.log("CLEARING");
          this.stage = INITIALIZING_STAGE_CLEARING;
        }
        break;
      case INITIALIZING_STAGE_CLEARING:
        if (this.clearingPattern.hasNext()) {
          const s = this.clearingPattern.next();
          step.setLinearMovement(s.getLinearMovement());
          step.setAngularMovement(s.getAngularMovement());
          if (s.getLinearMovement() == Model.Step.Movement.FORWARDS) {
            this.current++;
          }
        } else {
          console.log("TARGETING");
          this.stage = INITIALIZING_STAGE_TARGETING;
        }
        break;
      case INITIALIZING_STAGE_TARGETING:
        if (!this.targetingPattern.hasNext()) {
          console.log("DONE");
          this.complete = true;
          break;
        }
        const s = this.targetingPattern.next();
        step.setLinearMovement(s.getLinearMovement());
        step.setAngularMovement(s.getAngularMovement());
        if (s.getLinearMovement() == Model.Step.Movement.BACKWARDS) {
          this.current--;
        }
        if (this.current <= this.target) {
          console.log("DONE");
          this.complete = true;
        }
        break;
    }
    return step;
  }

  isExternalPattern() {
    return false;
  }
}

module.exports = InitializingPatternIterator;